hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1965,6 +2014,68 @@ public String listStatus(final Path path, final String startFrom,
     return continuation;
   }
 
+  private void addBlobListAsFileStatus(final BlobList blobList,
+      List<FileStatus> fileStatuses) throws IOException {
+
+    // Here before adding the data we might have to remove the duplicates.
+    // List Blobs call over blob endpoint returns two types of entries: Blob
+    // and BlobPrefix.  In the case where ABFS generated the data,
+    // there will be a marker blob for each "directory" created by driver,
+    // and we will receive them as a Blob.  If there are also files within this
+    // "directory", we will also receive a BlobPrefix.  To further
+    // complicate matters, the data may not be generated by ABFS Driver, in
+    // which case we may not have a marker blob for each "directory". In this
+    // the only way to know there is a directory is using BlobPrefix entry.
+    // So, sometimes we receive both a Blob and a BlobPrefix for directories,
+    // and sometimes we receive only BlobPrefix as directory. We remove duplicates
+    // but prefer Blob over BlobPrefix.
+    TreeMap<String, FileStatus> fileMetadata = new TreeMap<>();
+    List<BlobProperty> blobProperties = blobList.getBlobPropertyList();
+
+    for (BlobProperty entry: blobProperties) {
+      String blobKey = entry.getName();
```

Comment:
Can this be added as a separate function, too big a code piece at a place and if possible same function can handle both blob and dfs based on some intelligence ?